File: Bomba_Combustivel.rb
class Bomba_Combustivel

  def initialize(capacidade, preco_por_litro)
    validar_bomba(capacidade, preco_por_litro)
    @capacidade = capacidade
    @preco_por_litro = preco_por_litro
    @quantidade_presente_na_bomba = 0
  end

  def encher()
    @quantidade_presente_na_bomba = @capacidade
  end

  def abastecer_por_valor(valor)
    quantidade_abastecida = valor/@preco_por_litro
    validar_abastecimento(quantidade_abastecida)
    @quantidade_presente_na_bomba -= quantidade_abastecida
    quantidade_abastecida
  end

  def abastecer_por_litros(quantidade_litros)
    validar_abastecimento(quantidade_litros)
    @quantidade_presente_na_bomba -= quantidade_litros
    quantidade_litros * @preco_por_litro
  end

  def alterar_preco(novo_preco)
    @preco_por_litro = novo_preco
  end

  def retornar_capacidade()
    @capacidade
  end

  def retornar_preco_por_litro()
    @preco_por_litro
  end

  def retornar_quantidade_combustivel()
    @quantidade_presente_na_bomba
  end

private

  def validar_bomba(capacidade, preco_por_litro)
    raise if capacidade <= 0 || preco_por_litro <= 0
  end

  def validar_abastecimento(quantidade_litros)
    raise if quantidade_litros > @quantidade_presente_na_bomba
  end
end

File: BombaCombustivel.java

public class BombaCombustivel {

	private int capacidade;
	private double precoPorLitro;
	private boolean cheia;
	private double quantidadeCombutivelPresente;

	public BombaCombustivel(int capacidade, double precoPorLitro) {
		validarCapacidade(capacidade);
		validarPreco(precoPorLitro);
		this.capacidade = capacidade;
		this.precoPorLitro = precoPorLitro;
		this.cheia = false;
		this.quantidadeCombutivelPresente = 0;
	}
	
	public void encher(){
		this.quantidadeCombutivelPresente = this.capacidade;
		this.cheia = true;
	}
	
	public boolean cheia(){
		return this.cheia;
	}

	public double quantidadeCombustivelPresente() {
		return this.quantidadeCombutivelPresente;
	}
	
	public int capacidade(){
		return this.capacidade;
	}
	
	public double precoPorLitro(){
		return this.precoPorLitro;
	}

	public double abastecerPorValor(double valor) {
		validarAbastecimentoPorValor(valor);
		double litrosAbastecidos = valor/this.precoPorLitro;
		this.quantidadeCombutivelPresente -= litrosAbastecidos;
		return litrosAbastecidos;
	}

	public double abastecerPorQuantidadeEmLitros(double quantidade) {
		validarAbastecimentoPorQuantidade(quantidade);
		this.quantidadeCombutivelPresente -= quantidade;
		return quantidade*this.precoPorLitro;
	}

	public void alterarPreco(double novoPreco) {
		validarPreco(novoPreco);
		this.precoPorLitro = novoPreco;		
	}
	
	private void validarAbastecimentoPorQuantidade(double quantidade){
		if (quantidade > this.quantidadeCombutivelPresente)
			throw new RuntimeException("Bomba nao tem combustivel suficiente para o abastecimento!");
	}
	
	private void validarAbastecimentoPorValor(double valor){
		if (valor > this.quantidadeCombutivelPresente*this.precoPorLitro)
			throw new RuntimeException("Bomba nao tem combustivel suficiente para o abastecimento!");
	}
	
	private void validarCapacidade(int capacidade){
		if (capacidade <= 0)
			throw new RuntimeException("Capacidade Invalida!");
	}
	
	private void validarPreco(double preco){
		if (preco <= 0)
			throw new RuntimeException("Preco Invalido!");
	}

}
